### PR TITLE
Parameters enhancements

### LINKED ISSUE #4653: Preview of embed breaks with x'd out default value
to repro 

SQL
```
select count(*) from orders
where 1=1 [[AND {{ts}}]]
```
the `ts` parameter should be a field filter  linked to "created_at" in orders, with a default value of today.

If I open the embedding preview, and enable `ts`, then 'x' out the default, I get the below error

screen shot 2017-03-30 at 2 11 58 pm

```

:query
{:type "native",
:native
{:query "select count(*) from orders\nwhere 1=1 [[AND {{ts}}]]",
:collection "ORDERS",
:template_tags {:ts {:id "30af3240-fd9f-d1d1-4762-9c0e1eefcfdc", :name "ts", :display_name "Ts", :type "dimension", :dimension ["field-id" 1], :widget_type "date/relative", :default "today"}}},
:parameters ({:type "date/relative", :target ["dimension" ["template-tag" "ts"]], :value ""}),
:constraints {:max-results 10000, :max-results-bare-rows 2000},
:info
{:executed-by 1,
:context :embedded-question,
:card-id 4,
:dashboard-id nil,
:query-hash [27, -62, -88, -127, -20, 6, 16, -84, 56, -13, 5, 52, -112, -43, -104, -49, -89, 59, -66, -69, 115, 103, 30, 48, -97, -52, 39, 119, -24, 110, 42, -100],
:query-type "native"}},
:expanded-query nil}

Mar 30 14:11:41 WARN metabase.query-processor :: Query failure: Invalid format: ""
["query_processor$assert_query_status_successful.invokeStatic(query_processor.clj:149)"
"query_processor$assert_query_status_successful.invoke(query_processor.clj:142)"
"query_processor$run_and_save_query_BANG_.invokeStatic(query_processor.clj:181)"
"query_processor$run_and_save_query_BANG_.invoke(query_processor.clj:175)"
"query_processor$eval35779$dataset_query__35784$fn__35785.invoke(query_processor.clj:212)"
"query_processor$eval35779$dataset_query__35784.invoke(query_processor.clj:199)"
"api.card$run_query_for_card.invokeStatic(card.clj:373)"
"api.card$run_query_for_card.doInvoke(card.clj:357)"
"api.public$run_query_for_card_with_id$fn__39433.invoke(public.clj:59)"
"api.public$run_query_for_card_with_id.invokeStatic(public.clj:57)"
"api.public$run_query_for_card_with_id.doInvoke(public.clj:53)"
"api.embed$run_query_for_card_with_params.invokeStatic(embed.clj:196)"
"api.embed$run_query_for_card_with_params.doInvoke(embed.clj:188)"
"api.preview_embed$fn__40180$fn__40181.invoke(preview_embed.clj:32)"
"api.common.internal$do_with_caught_api_exceptions.invokeStatic(internal.clj:229)"
"api.common.internal$do_with_caught_api_exceptions.invoke(internal.clj:224)"
"api.preview_embed$fn__40180.invokeStatic(preview_embed.clj:27)"
"api.preview_embed$fn__40180.invoke(preview_embed.clj:27)"
"middleware$enforce_authentication$fn__36253.invoke(middleware.clj:119)"
"api.routes$fn__41063.invokeStatic(routes.clj:70)"
"api.routes$fn__41063.invoke(routes.clj:70)"
"routes$fn__41117.invokeStatic(routes.clj:44)"
"routes$fn__41117.invoke(routes.clj:44)"
"middleware$log_api_call$fn__36368$fn__36370.invoke(middleware.clj:331)"
"middleware$log_api_call$fn__36368.invoke(middleware.clj:330)"
"middleware$add_security_headers$fn__36302.invoke(middleware.clj:246)"
"middleware$bind_current_user$fn__36257.invoke(middleware.clj:139)"
"middleware$maybe_set_site_url$fn__36306.invoke(middleware.clj:268)"]
```

### LINKED ISSUE #4655: Enable FieldValues filtering in SQL Template Filter Widgets
These were not included in #4646 , but should be included

### LINKED ISSUE #4663: Should parameter widgets work on public questions?
I'm not sure if there are good reasons why we're not doing this, but if you have a filter widget mapped to a variable in a sql question and then you share it as a public link, the widget(s) is not exposed: 

---
The query:

screen shot 2017-03-31 at 11 43 36 am

----

The public link:

screen shot 2017-03-31 at 11 43 51 am

### PR DESCRIPTION
**Pull request by tlrobinson**
- [x] field values support in question parameters. Resolves #4655
  - [x] If we want this to work on public/embeded questions the query endpoints need to return `param_values` like for dashboards.
- [x] fix bug when clearing embedded parameter Resolves #4653
- [x] parameter widgets on public questions Resolves #4663

### CODE REVIEW

**Comment by salsakran** [comment]
👍  on moving param code out of the dashboard components

**Comment by tlrobinson** [comment]
@salsakran @camsaul thoughts on the remaining TODO above? It will need a little backend work to return `param_values` like we do for dashboards.

**Comment by salsakran** [comment]
I think it's important that widgets work the same regardless of where they are. 

I also think that making end users manually enter strings for parameters is completely un-zen. Any situation where FieldValues are available we should use them.

**Comment by tlrobinson** [comment]
Ok, @camsaul can you handle that on the backend?

There's a little frontend work as well but I can do that once the backend is done.

**Comment by camsaul** [comment]
yeah, i'll take a look at it 

Right now trying to get Druid 0.9.2 set up for CI is giving me a huge headache so it will be a while.

File: frontend/src/metabase/dashboard/selectors.js [hunk 1]
```diff
@@ -143,6 +146,23 @@ export const getMappingsByParameter = createSelector(
     }
 );
 
+export const getParameters = createSelector(
```
Comment (attekei) [inline frontend/src/metabase/dashboard/selectors.js]: Would be nice to have a comment about the behavior

File: frontend/src/metabase/lib/query/field.js [hunk 2]
```diff
@@ -21,26 +22,38 @@ export function getFieldTargetId(field: Field): ?FieldId {
...
     return Array.isArray(field) && field.length === 2 && mbqlEq(field[0], "expression");
 }
 
-export function isAggregateField(field: Field): boolean {
+export function isAggregateField(field: FieldReference): boolean {
     return Array.isArray(field) && mbqlEq(field[0], "aggregation");
 }
+
+// https://github.com/metabase/metabase/issues/3417
+export function getFieldValues(field: ?Field): any[] {
```
Comment (attekei) [inline frontend/src/metabase/lib/query/field.js]: Would be nice to have a comment about the behavior, not only a link to GitHub issue

File: frontend/src/metabase/meta/Parameter.js [hunk 3]
```diff
@@ -25,9 +30,26 @@ export function getTemplateTagParameters(tags: TemplateTag[]): Parameter[] {
...
     for (const parameter of parameters) {
-        if (parameterValues[parameter.id] !== undefined) {
+        if (parameterValues[parameter.id] != undefined) {
             result[parameter.slug] = parameterValues[parameter.id];
         }
     }
     return result;
 }
+
+export function getParameterTargetFieldId(target: ?ParameterTarget, datasetQuery: DatasetQuery): ?FieldId {
```
Comment (attekei) [inline frontend/src/metabase/meta/Parameter.js]: Again, rather hard-to-read logic so short comment explaining the behavior would be nice

**Review by attekei — CHANGES_REQUESTED** [review]
After working for a while with the backend I realized that the parts of frontend involving heavy business logic should ideally have comments and/or examples demonstrating how they work. The comments above are on methods I found most obscure to me.

SQL parameters in query builder seem to be broken in the current state of this PR.

**Comment by attekei** [comment]
Also typing with Flow would have a really nice self-documenting effect – even if you have to use many `// $FlowFixMe` annotations  that isn't IMHO that big offence.

**Comment by tlrobinson** [comment]
In general I add Flow types to new files and smaller existing files when modifying them, but big ones can take a lot of time so I don't always do them. I try to avoid `$FlowFixMe` but I suppose it's not as bad as using `any` types so maybe I'll be more aggressive about adding Flow types.

**Comment by tlrobinson** [comment]
@camsaul Are you able to do the backend portion of this soon? @attekei has a branch he'd like to merge with this one sooner rather than later.

**Comment by camsaul** [comment]
@tlrobinson what specific endpoint(s) do you need

**Comment by tlrobinson** [comment]
@camsaul embed/public link card API endpoints, like we have for `/api/public/dashboard/:uuid`. So `/api/public/card/:uuid` and `/api/embed/card/:token` I think.

**Comment by camsaul** [comment]
ooook

**Comment by camsaul** [comment]
@tlrobinson do you know where should these params come from? For dashboards we just look at the values of `:parameter_mappings` for each DashboardCard, but Cards don't have anything like that

**Comment by tlrobinson** [comment]
There's a function called `template-tag-parameters` and `add-implicit-card-parameters` in `metabase/api/embed.clj` that generates the parameters from the SQL `template_tags`. Currently only used on `GET /api/embed/card/:id` I believe.

**Comment by camsaul** [comment]
@tlrobinson there is no `GET /api/embed/card/:id` endpoint. `GET /api/embed/card/:token` is already calling `add-implict-card-parameters`, so I'm guessing you only need that added to `/api/public/card/:uuid`, right?

**Comment by tlrobinson** [comment]
@camsaul I need `param_values` like we have for dashboards.

**Comment by camsaul** [comment]
ok @tlrobinson changes are done let me know if they work for you

**Comment by tlrobinson** [comment]
I think this is ready, but might be a good candidate for an E2E test

**Comment by salsakran** [comment]
tests would be nifty.

**Comment by salsakran** [comment]
can we un-muck the section that papers over #3417 as part of this?

**Review by salsakran — APPROVED** [review]
Overall, looks reasonable. The parameters code in general is pretty dense and hard to follow, but this doesn't regress in any way.

Some frontend tests for the parameters code would be ideal, but not a blocker.

**Review by attekei — APPROVED** [review]

**Comment by tlrobinson** [comment]
@attekei I got all the E2E tests passing except your 2 dashboard tests. I've disabled them for now so we can get stats and metabase-head building/updating again (its been 2 weeks). Feel free to fix those and re-enable them at your convenience.
